### PR TITLE
gateway refactoring

### PR DESCRIPTION
**Pull request by tmcgroul**

### CODE REVIEW

**Comment by tmcgroul** [comment]
@ozgrakkurt
a few points that i wanted to achive:
1. remove different implementations for postgres and cockroach and write a code which will work in both databases correctly
2. do not use json fucntions in sql because it makes sql queries less readable
3. make the code easy to understand:)

but it feature incomplete in comparison with current implementation and still seems difficult to understand
so i don't want to merge it now, but i am going to take some ideas from this pr and implement in current codebase... but later and step by step without breaking changes

File: src/archive/fields.rs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+            _all: value,
+            index_in_block: value,
+            signature: value,
+            success: value,
+            hash: value,
+            call: CallFields::new(value),
+        }
+    }
+
+    pub fn any(&self) -> bool {
```
Comment (ozgrakkurt) [inline src/archive/fields.rs:153]: this might be better implemented like
```
return self.a || self.b || self.c
```

File: src/archive/fields.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        }
+        if self.call.any() {
+            return true;
+        }
+        false
+    }
+
+    pub fn selected_fields(&self) -> Vec<&str> {
+        let mut fields = vec![];
+        if self._all {
```
Comment (ozgrakkurt) [inline src/archive/fields.rs:177]: can do fields.extend_from_slice(&[a, b, c]) maybe

File: src/archive/fields.rs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+                fields.push("call_id");
+            }
+        }
+        fields
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct EventFields {
+    pub _all: bool,
```
Comment (ozgrakkurt) [inline src/archive/fields.rs:206]: pretty sure this doesn't have to be _all it can just be all

**Comment by ozgrakkurt** [comment]
@tmcgroul added some comments about rust code

File: src/archive/fields.rs (hunk 2 above)
Comment (tmcgroul) [reply src/archive/fields.rs:177]: haven't heard about `extend_from_slice`
i was sure that there is should be a more efficient way than multiple push calls
will change to this method

File: src/archive/fields.rs (hunk 1 above)
Comment (tmcgroul) [reply src/archive/fields.rs:153]: yea
your suggestion looks cleaner, will change

**Comment by tmcgroul** [comment]
> @tmcgroul added some comments about rust code

implemented the suggestions in main branch

**Pull request closed without merging**